project/src/main/webapp/gmail.js

```diff
@@ -12,193 +12,206 @@
                             'userId': 'me',
-                            'id': result[i].id,
-                            'format': "raw"
+                            'pageToken': nextPageToken,
+                            'q': this.question
                         });
-                        var messageID = result[i].id;
-                        messageRequest.execute(response => {
-                            const decodedEmail = atob(response.raw.replace(/-/g, '+').replace(/_/g, '/')); 
-                            const emailBodyStartIndex = decodedEmail.indexOf("Content-Type: text/plain; charset=\"UTF-8\"");
-                            const emailBodyEndIndex = (decodedEmail.substring(emailBodyStartIndex)).indexOf("Content-Type: text/html; charset=\"UTF-8\"")
-                            const emailBodyValue = decodedEmail.substring(emailBodyStartIndex, emailBodyStartIndex+emailBodyEndIndex);
-                            const emailDateStartIndex = decodedEmail.indexOf("Date:");
-                            const emailDateEndIndex = decodedEmail.substring(emailDateStartIndex).indexOf("Message-ID");
-                            var emailDateValue = decodedEmail.substring(emailDateStartIndex, emailDateStartIndex+emailDateEndIndex);
-                            if (emailDateValue.indexOf("X-Notifications") != -1) {
-                                emailDateValue = emailDateValue.substring(0, emailDateValue.indexOf("X-Notifications"));
-                            }
-                            var emailObject = {};
-                            emailObject[messageID] = {emailDate: emailDateValue, emailBody: emailBodyValue}; 
-                            if (emailBodyValue.indexOf("format=flowed") == -1 && emailDateValue.length > 1) {
-                                list[list.length] = emailObject;
+                        getPageOfMessages(request, result);
+                    } else {
+                        for (var i = 0; i < result.length; i++) {
+                            if (this.hasEmails(result)) {
+                                promiseArray.push(this.getMessage(result[i].id));
                             }
-                        });
-                    }
+                        }
+                        Promise.all(promiseArray).then(() => resolve()
+                        );
                     }
                 });
             };
-            console.log(list);
             var initialRequest = gapi.client.gmail.users.messages.list({
-            'userId': 'me',
-            'q': "When is the interview prep workshop?"
+                'userId': 'me',
+                'q': this.question
+            });
+            getPageOfMessages(initialRequest, []);
+            return Promise.all(promiseArray);
         });
-        getPageOfMessages(initialRequest, []);
-        const t1 = performance.now();
-        console.log(t1-t0);
+        return listMessagePromise;
```

Comment:
This can be combined with line 141 by just returning the promise right away:
```
return new Promise((resolve) => {
```

Reply:
done!